WIP Integrate Swagger UI

feorlen opened the pull request:
PR notes: this version should work, for anyone who would like to try it out. Both Swagger UI and django-cors-headers recently released new versions, so I will test those and update accordingly. If you would like to see a working instance, I can set one up on my droplet. The issue for this PR is https://github.com/zulip/zulip/issues/3474

Generate REST API documentation, with working examples.

New directories:
* static/swagger/
* static/yaml/

New dependancies:
* Swagger UI 2.2.10
* django-cors-headers 2.0.0

Configuration:
* yaml/zulip.yaml: Zulip API definitions
* zproject/settings.py: configuration for django-cors-headers

The new documentation is available in Zulip at /apidocs/

Swagger UI is an open source package that provides a graphical front-end
for REST API documentation, including examples against a live backend. The
API is described in a YAML file using the format defined by the Swagger/OpenAPI
specification.

For Zulip, I have integrated Swagger UI in its simplest form by using the
dist directory provided in the github project, putting its contents in a new
static/swagger directory in the Zulip directory structure, and creating a new
url in zproject/urls.py.

The API definition file, zulip.yaml, is in a new static/yaml directory to keep
it separate from files that come from the Swagger UI project. This first
version documents the /messages endpoint. Additional endpoint documentation
can be added by including new definitions in zulip.yaml.

Note: There is not currently an automated way to update the dist files
from the Swagger UI project. To do this, replace the contents of the
static/swagger directory with new dist contents from the Swagger UI project.

Swagger UI requires the target backend support Cross-Origin Resource Sharing
(CORS) headers. To do this, I have added django-cors-headers as a dependancy
(with its configuration) in zproject/settings.py.

This change has been tested on a clean host with a new Zulip installation from
my feature branch. The new doc UI is available at /apidocs/, which redirects
to /static/swagger/index.html?url=/static/yaml/zulip.yaml.

How to access the new docs from the Zulip UI has yet to be decided. No Zulip
UI changes are included here.

References:

* Swagger UI and the Swagger/OpenAPI spec:
  * http://swagger.io
* Swagger UI github project:
  * https://github.com/swagger-api/swagger-ui
* django-cors-headers github project:
  * https://github.com/ottoyiu/django-cors-headers

smarx commented:
*Automated message from Dropbox CLA bot*

@feorlen, it looks like you've already signed the Dropbox CLA. Thanks!

feorlen commented:
This is passing tests now, @timabbott let me know what you think and if there's anything that needs to be addressed.

feorlen commented:
The current version is running on my droplet, feorlen.zulipdev.org with the usual port. Zoe already has a bot created for testing so you can use that login if you like.

Create your test bot, go to /apidocs/ and authenticate with the button at the top of the page. (I usually start from the existing API documentation page from the gear menu, then change the url.) Then you can send messages via the /messages endpoint.

feorlen commented:
@timabbott , I don't know anything about npm, can it install only part of a package? The files I'm using are one directory, not the whole thing. I guess it couldn't harm anything by installing all of it, but there's a lot more and it's only needed if we were customizing the UI.

feorlen commented:
I tried to see if npm could install only part of a package, but I didn't find anything that suggests packages aren't atomic units. Also npm only installs in its single global directory.

The `dist` directory has to be accessible at a url so its index.html (and other css/js) can be loaded. By "changing some paths" do you mean using a redirect (or the equivalent of Apache Alias) from `/apidocs/` to have it load from wherever npm installs things? 

The full swagger-ui package includes a nginx.conf, so that must be how it serves content from `dist` at the default http://localhost:8080 location for a custom-built install. There are also javascript and node dependencies. (See https://github.com/swagger-api/swagger-ui#customize)

I've posted a question to the mailing list asking how people are managing upgrades. I don't have an answer right now besides manually updating the contents of `dist`. Several different docs and other comments from the package maintainers talk about just taking that directory and putting it in an appropriate location (on a webserver.) 

From a sysadmin standpoint, I'd be concerned about web content being served from an arbitrary place on the filesystem. I'm not sure how else it would work if npm can only install packages in a single location. I found a few people asking if they could have a "local" npm installation location separate from the rest, and several comments from the project saying they would never consider it.

feorlen commented:
One of the regulars on the mailing list reports that some people are using a script to move the desired files into place from the install location. Is that possible?

feorlen commented:
I've removed the Swagger UI files and added the package as an npm dependency. `node_cache.py` now copies the dist directory into `zulip/static`. I've tested this on a clean host and everything still works as expected.

I've left this as a separate commit for the moment to make the new changes easier to see. @timabbott I'd appreciate any comments you have on this approach.

feorlen commented:
If things have calmed down in PR-mergeville hopefully I can move forward with the Swagger changes. @timabbott is there someone other than yourself who might be a good person to review this? It would be helpful to get some advice on if I'm doing the npm part correctly.

tfheen commented:
This looks very good to me. (I'm @feorlen's Outreachy mentor.)

feorlen commented:
Hi @timabbott, is there something I can add to make this easier to evaluate? I've also still got the test host if that would be helpful to start up again, I know this is a large change. When you get some time, I'd appreciate any recommendations you might have. Thanks!

feorlen commented:
I just upgraded my prod install from this branch, and everything worked as expected.

feorlen commented:
Fyi, I'm adding to the yaml file to stub out currently undocumented return values for the endpoints documented in this first set. That will be done shortly. (I figured I had a little time while Travis issues got sorted out.)

timabbott commented:
So I just tried testing this out in development.  I notice that if you're logged in, you get a prompt for username/password when you try to use it; I think that's because we're posting to `/api/v1` type URLs (which require HTTP basic auth), not `/json` URLs (which would use your cookie).  I think if we can make Swagger do this, it'd be awesome to post to the `/json` URLs so that testing with the API from the browser "just works".  

Is that doable?  

Another thing that would be nice is to have the "default" block be expanded to show the list of endpoints when you land on the page.

I also notice that raw `http://127.0.0.1:9991/static/swagger/index.html` has totally different content involving a petstore (presumably whatever comes with it); is there a way we can set the default for Swagger to be our `zulip.yaml` file?

timabbott commented:
Also, what's the story with this sample response for the message posting URL?  It seems to not correspond to our API...

![image](https://cloud.githubusercontent.com/assets/2746074/23586032/d97e7192-0141-11e7-937b-7b9bffdc5407.png)

feorlen commented:
Handling these one at a time...

Going to Swagger's index.html directly gets the default version, which will send you to their petstore sample. You can provide a url of the yaml file you want when you load that page, which is how I'm currently avoiding having a custom version of it. It can be a symlink to an actual custom version elsewhere, but something needs to be in that directory so Swagger UI can find the rest of its files. 

So we could have a custom index.html (in static/html? Could have a different name, too) and then create the link in static/swagger. Presumably the same way I'm currently copying files over from the installed package.

feorlen commented:
The "default" response is from the sample yaml file, anything could be there instead. The 200 response is shown at the top as the success case, and the others below. For example, here I added an additional definition for 400: 

![screen shot 2017-03-07 at 13 04 04](https://cloud.githubusercontent.com/assets/95028/23677890/b9827568-0336-11e7-8b6e-6b39ddd100c5.png)

feorlen commented:
I can control how much is displayed when the page loads by changing the value of docExpansion in index.html. The options are `none` (what you see), `list` (shows endpoints), and `full` (shows everything.) This is `list`:

![screen shot 2017-03-07 at 13 20 06](https://cloud.githubusercontent.com/assets/95028/23678391/cf08e78a-0338-11e7-803a-08fd03a00f91.png)

feorlen commented:
For auth, the standard options are Basic, API key, or OAuth2. Is what you mean with the cookie like what some people have done by adding a custom header with a CSRF token? When I tried /json endpoints with Basic auth, I got a CSRF error, so I thought it was related.

See https://github.com/swagger-api/swagger-ui/issues/1380. I don't know if the custom code is in index.html or somewhere else, the example referenced in that issue mentions a file index.html.js and nothing with that name exists. I haven't tried to make this work yet.

timabbott commented:
Re expansion, I think  `list` is the right choice, since I imagine we'll have 50 of these eventually.

Copying over a replacement index.html is probably reasonable.  

For JSON endpoints, you don't need to send any auth stuff at all, as long as the user is logged-in (has a valid cookie).  Basically, if the user is doing API work not on the website, they want to use the /api/v1 endpoints and use our API auth.  But if they're on the website, they're probably logged in, and we should just use their existing browser session cookie (by sending to the `/json` endpoint).  Hopefully swagger supports posting to a different URL than advertised...

feorlen commented:
The other items are straightforward, I'll have a new version to try next week.

/api/v1 vs /json could be too much magic. There have been folks with multiple APIs (basePaths), but they wanted to display all of them in one UI. I don't see anything that allows one in certain circumstances and another in others. Swagger UI expects to function just like any other API user. (It could be on an entirely different host and webserver, for example.)

All this can be added to the text descriptions, so at least it's clear what the alternatives are.

timabbott commented:
I don't think we actually want anyone using the /json API except for the webapp and the Swagger tool.  I'm just thinking that if we do the trick I suggested, users wouldn't have to enter their API key to test things, which would be really slick.

feorlen commented:
Moving this back to WIP for now. 

There is a new version of Swagger UI (as of a few days ago) that may allow some of the things we'd like to have, but seems to work differently. Copying over one directory of files doesn't seem to work anymore. I'm trying to find out what the new version can and can't do, the project team has said they won't support the old one for more than a few months.

feorlen commented:
I'm trying to get the new Swagger UI to work like the old one, it is apparently a major re-write. The documentation has conflicting information about how to use the dist directory to serve generated API docs. 

The currently checked in code works, but still uses the old version. I will pick this up again next week when I'm back online. 

Also, Akash Kothawale expressed interest in working on Swagger API documentation. (If you see this, may I get your github id?)

Here is what I know so far about the newest version of Swagger UI, [3.0.2](https://github.com/swagger-api/swagger-ui/tree/v3.0.2):

* When I install with npm, most of the files that were previously in node_modules/swagger-ui/dist are not there. The new swagger-ui/public directory, where the index.html is supposed to be, is missing. This file is a replacement for /dist/index.html (which is still there, but doesn't work.)

* There is an additional new thing installed in node_modules, swagger-client. I can't find any reference to what this is, but swagger-ui appears to depend on it (at least for some configurations.)

* This version says it supports only OpenAPI spec version 2.0, which is what zulip.yaml uses.

* When I manually copy files from a downloaded tarball (using default /dist and /public directory names since now it has hardcoded paths) the UI loads but no longer displays a target yaml file provided in the url. 

* If I type the url to zulip.yaml into the ui, it says "No operations defined in spec!". It appears to load the file (and says it's valid) but doesn't display anything.

I have some recent conversations on the Swagger mailing list, hopefully someone from the project will have recommendations. They appear to be still working on documenting the new way of doing things.

[no dist directory install anymore?](https://groups.google.com/forum/#!topic/swagger-swaggersocket/I7OwoblAB0I)
[how do do dist dir install of swagger-ui 3.0.2?](https://groups.google.com/forum/#!topic/swagger-swaggersocket/JKuP1u_07s8)

feorlen commented:
It looks like Swagger UI 3.0.5 works as expected. I'm going to do some further tests with a clean install to confirm.

fyi: @decached

feorlen commented:
Nope. Some files hanging around my dev environment were making it work. When I do a clean install, lots of necessary files are missing. Back to where we were.

feorlen commented:
I can take the dist directory from a downloaded tar and use that. But the npm package is broken.

I've posted another message to the mailing list. I've not seen a response to my earlier one on the same subject, perhaps this time someone will have a suggestion. https://groups.google.com/d/msg/swagger-swaggersocket/JKuP1u_07s8/5bTANDfMBgAJ

feorlen commented:
It looks like this was a design decision by the package maintainers, on the idea that anyone installing by npm is building the Swagger UI package themselves in a development environment. There are several people reporting problems. Here's one:

Plans for publishing package with built assets?
https://github.com/swagger-api/swagger-ui/issues/2758

feorlen commented:
I just updated this PR to use the new swagger-ui-dist npm module. It's working, but there's a problem that seems to be something in the new zulip.yaml file. My Zulip back end isn't getting requests from Swagger UI, yet the server definition hasn't changed. 

With the new stuff, does anyone even want Swagger UI anymore?

timabbott commented:
I think we may want Swagger UI, in addition to using something like readme.io that does this for us, yeah.

feorlen commented:
Isolated the cause of my trouble, there are two things: 
* The Swagger UI index.html apparently now requires the full url and not just a relative path to the yaml file. This is what is passed to its yaml validation service.
* That validation is unhappy with several things in the new zulip.yaml. If I replace it with my old file, I can correctly talk to the back end.

I'm investigating. 

Putting a real url in the file isn't going to work (whose domain?) But the only reason to do this is so if someone explicitly loads swaggerindex.html (without the redirect that specifies a yaml file) they will get the Zulip API and not the default Petstore. Maybe there's some new way to do a relative path. (I changed the name from index.html to swaggerindex.html so if someone finds it on their own there is at least a clue what it is.)

As for validation errors in the file, presumably that's something I can fix. I'll coordinate with @YagoGG to make sure it still works for everybody.

YagoGG commented:
I can confirm the validation issues, [readme.io](readme.io) also complained about them when I tried to re-sync.

@feorlen Feel free to take care of them if you have some spare time. Otherwise, I'll fix them as soon as I'm available (in a couple weeks).

YagoGG commented:
#5183 should fix the validation issues. Thanks for the heads up, @feorlen!

timabbott commented:
I'm going to close this, since I'm not sure we would want Swagger running on the same server as Zulip, and it's also very stale.

But I wanted to thank you for getting Zulip started towards OpenAPI @feorlen!  We now have 12K lines of OpenAPI data in `zulip.yaml` with full documentation for 60 endpoints.
